File: node/sample_05_balance.js
/**
 * Sample 5 — Lay so du tai khoan (Account Balance)
 * ===================================================
 * Kiem tra kha nang giao dich truoc khi dat lenh.
 *
 * Luong:
 *   1. Client goi endpoint balance theo accountNo
 *   2. API tra ve available, onHold, limits, settlement
 *   3. Tinh kha nang mua/ban thuc te theo nghiep vu
 *   4. Neu khong du dieu kien thi chan thao tac dat lenh
 *   5. Neu du dieu kien thi cho phep di tiep sang order flow
 */

import { Auth, Trading } from '@ssi.developer/ssi-sdk';
import { config, ACCOUNT_NO, OTP } from './config.js';
import { ensureAuth } from './auth_helper.js';

const vnd = (x) => Number(x).toLocaleString('en-US', { maximumFractionDigits: 0 });

const auth = new Auth(config);
await ensureAuth(auth, OTP);

const trading = new Trading(auth);

// --- Buoc 1: Lay danh sach tai khoan ---
const accounts = await trading.account.getAccountInfo();
console.log('Danh sach tai khoan:');
for (const acc of accounts) {
  console.log(`  - ${acc.accountNo} (${acc.accountType})`);
}

// --- Buoc 2: Lay so du tai khoan Equity ---
console.log(`\n--- So du tai khoan Equity: ${ACCOUNT_NO} ---`);
const balance = await trading.portfolio.getEquityBalance(ACCOUNT_NO);
console.log(`  Tien mat kha dung : ${vnd(balance.availableCash).padStart(15)}`);
console.log(`  Tong no           : ${vnd(balance.totalDebt).padStart(15)}`);
console.log(`  Mua T0/T1/T2      : ${vnd(balance.buyT0).padStart(12)} / ${vnd(balance.buyT1).padStart(12)} / ${vnd(balance.buyT2).padStart(12)}`);
console.log(`  Ban T0/T1/T2      : ${vnd(balance.sellT0).padStart(12)} / ${vnd(balance.sellT1).padStart(12)} / ${vnd(balance.sellT2).padStart(12)}`);

// --- Buoc 3: Kiem tra suc mua toi da cho mot ma ---
console.log('\n--- Suc mua/ban toi da: SSI ---');
const maxBs = await trading.trading.getMaxBuySell(ACCOUNT_NO, 'SSI', 26000);
console.log(`  Max mua : ${String(maxBs.maxBuyQuantity).padStart(10)} co phieu`);
console.log(`  Max ban : ${String(maxBs.maxSellQuantity).padStart(10)} co phieu`);
console.log(`  Suc mua : ${maxBs.purchasePower}`);

// --- Buoc 4: Logic kiem tra truoc khi dat lenh ---
const desiredQuantity = 100;
const desiredPrice = 26000;
const requiredAmount = desiredQuantity * desiredPrice;

if (balance.availableCash >= requiredAmount) {
  console.log(`\n[OK] Du dieu kien: can ${vnd(requiredAmount)}, co ${vnd(balance.availableCash)}`);
  console.log('  -> Cho phep dat lenh mua.');
} else {
  console.log(`\n[X] Khong du: can ${vnd(requiredAmount)}, chi co ${vnd(balance.availableCash)}`);
  console.log('  -> Chan dat lenh.');
}

// --- Buoc 5: Xem vi the hien co ---
console.log(`\n--- Vi the co phieu (${ACCOUNT_NO}) ---`);
const positions = await trading.portfolio.getEquityPositions(ACCOUNT_NO);
for (const pos of positions) {
  console.log(
    `  ${pos.symbol.padEnd(10)} | SL: ${String(pos.quantity).padStart(8)} | ` +
    `Ban duoc: ${String(pos.sellableQuantity).padStart(8)} | Gia von: ${vnd(pos.costPrice).padStart(10)}`,
  );
}

// --- Response Summary ---
console.log(`\n[Response] accounts|avail_cash|max_buy_qty|max_sell_qty|positions`);
console.log(`${accounts.length}|${balance.availableCash}|${maxBs.maxBuyQuantity}|${maxBs.maxSellQuantity}|${positions.length}`);
if (positions.length > 0) {
  const p = positions[0];
  console.log(`[Response:first_pos] symbol|quantity|sellable|cost_price`);
  console.log(`${p.symbol}|${p.quantity}|${p.sellableQuantity}|${p.costPrice}`);
}

File: dotnet/Sample05Balance.cs
// Sample 05 — Lấy số dư tài khoản (Account Balance)
// Kiem tra kha nang giao dich truoc khi dat lenh.

using SsiSdk;

public static class Sample05Balance
{
    public static async Task RunAsync()
    {
        var config = SampleConfig.Create();
        using var auth = new AuthClient(config);
        await AuthHelper.EnsureAuthAsync(auth, SampleConfig.Otp);

        var trading = new TradingClient(auth);

        // --- Buoc 1: Lay danh sach tai khoan ---
        var accounts = await trading.Account.GetAccountInfoAsync();
        Console.WriteLine("Danh sach tai khoan:");
        foreach (var acc in accounts)
        {
            Console.WriteLine($"  - {acc.AccountNo} ({acc.AccountType})");
        }

        // --- Buoc 2: Lay so du tai khoan Equity ---
        Console.WriteLine($"\n--- So du tai khoan Equity: {SampleConfig.AccountNo} ---");
        var balance = await trading.Portfolio.GetEquityBalanceAsync(SampleConfig.AccountNo);
        if (balance is not null)
        {
            Console.WriteLine($"  Tien mat kha dung  : {balance.AccountBalance,15:N0}");
            Console.WriteLine($"  Tong no            : {balance.TotalDebt,15:N0}");
            Console.WriteLine($"  Mua T0/T1/T2       : {balance.BuyT0,12:N0} / {balance.BuyT1,12:N0} / {balance.BuyT2,12:N0}");
            Console.WriteLine($"  Ban T0/T1/T2       : {balance.SellT0,12:N0} / {balance.SellT1,12:N0} / {balance.SellT2,12:N0}");
        }

        // --- Buoc 3: Kiem tra suc mua toi da cho mot ma ---
        Console.WriteLine("\n--- Suc mua/ban toi da: SSI ---");
        var maxBs = await trading.Trading.GetMaxBuySellAsync(SampleConfig.AccountNo, "SSI", 26000);
        Console.WriteLine($"  Max mua : {maxBs.MaxBuyQuantity,10} co phieu");
        Console.WriteLine($"  Max ban : {maxBs.MaxSellQuantity,10} co phieu");
        Console.WriteLine($"  Suc mua : {maxBs.PurchasePower,15}");

        // --- Buoc 4: Logic kiem tra truoc khi dat lenh ---
        var desiredQuantity = 100;
        var desiredPrice = 26000.0;
        var requiredAmount = desiredQuantity * desiredPrice;

        if (balance is not null && balance.AccountBalance >= requiredAmount)
        {
            Console.WriteLine($"\n  Du dieu kien: can {requiredAmount:N0}, co {balance.AccountBalance:N0}");
            Console.WriteLine("  -> Cho phep dat lenh mua.");
        }
        else
        {
            Console.WriteLine($"\n  Khong du: can {requiredAmount:N0}, chi co {balance?.AccountBalance:N0}");
            Console.WriteLine("  -> Chan dat lenh.");
        }

        // --- Buoc 5: Xem vi the hien co ---
        Console.WriteLine($"\n--- Vi the co phieu ({SampleConfig.AccountNo}) ---");
        var positions = await trading.Portfolio.GetEquityPositionsAsync(SampleConfig.AccountNo);
        foreach (var pos in positions)
        {
            Console.WriteLine(
                $"  {pos.Symbol,-10} | SL: {pos.Quantity,8} | " +
                $"Ban duoc: {pos.SellableQuantity,8} | Gia von: {pos.CostPrice,10:N0}");
        }

        // --- Response Summary ---
        Console.WriteLine("\n[Response] accounts|avail_cash|max_buy_qty|max_sell_qty|positions");
        Console.WriteLine($"{accounts.Count}|{balance?.AccountBalance}|{maxBs.MaxBuyQuantity}|{maxBs.MaxSellQuantity}|{positions.Count}");
        if (positions.Count > 0)
        {
            var p = positions[0];
            Console.WriteLine("[Response:first_pos] symbol|quantity|sellable|cost_price");
            Console.WriteLine($"{p.Symbol}|{p.Quantity}|{p.SellableQuantity}|{p.CostPrice}");
        }
    }
}
